clippy_lints/src/instant_elapsed.rs

```diff
@@ -0,0 +1,90 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use rustc_errors::Applicability;
+use rustc_hir::{BinOpKind, BindingAnnotation, Expr, ExprKind, Node, PatKind, QPath, StmtKind, TyKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::source_map::Spanned;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Catch for substraction between Instant::new() and previous timestamp.
+    ///
+    /// ### Why is this bad?
+    /// This is error prone and there is a specific function for this.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::time::Instant;
+    /// let prev_instant = Instant::now();
+    /// let duration = Instant::now() - prev_instant;
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::time::Instant;
+    /// let prev_instant = Instant::now();
+    /// let duration = prev_instant.elapsed();
+    /// ```
+    #[clippy::version = "1.64.0"]
+    pub INSTANT_ELAPSED,
+    pedantic,
+    "catch substraction between Instant::new() and previous timestamp "
+}
+
+#[derive(Default)]
+pub struct InstantElapsed {
+    variable_name: String,
+    local_id: u32,
+}
+
+impl_lint_pass!(InstantElapsed => [INSTANT_ELAPSED]);
+
+impl LateLintPass<'_> for InstantElapsed {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ Expr<'_>) {
+        if let ExprKind::Block(block, _) = expr.kind {
+            for (_, stmt) in block.stmts.iter().enumerate() {
+                if let StmtKind::Local(local) = stmt.kind
+                    && let PatKind::Binding(BindingAnnotation::Unannotated, local_hir_id, ident, None) = local.pat.kind
+                    && let Some(expr_block) = local.init
+                    && check_instant_now_call(expr_block) {
+                    self.variable_name = ident.to_string();
+                    self.local_id = local_hir_id.local_id.as_u32();
+                }
+
+                if let StmtKind::Local(local) = stmt.kind
+                    && let Some(expr_block) = local.init
+                    && let ExprKind::Binary(Spanned {node: BinOpKind::Sub, ..}, lhs, rhs) = &expr_block.kind
+                    && check_instant_now_call(lhs)
+                    && let ExprKind::Path(qpath) = &rhs.kind
+                    && let QPath::Resolved(None, rhs_path) = qpath
+                    && let rustc_hir::def::Res::Local(expr_res) = rhs_path.res
+                    && expr_res.local_id.as_u32() == self.local_id {
```

Comment:
We don't need to track the creation of the RHS, as we can lint for any `Instant::now() - rhs` where `rhs` is an `Instant`. To check that you can use `cx.typeck_results().expr_ty(rhs)`

To make the suggestion, `Sugg::hir_opt(cx, rhs).maybe_par()` (or `Sugg::hir_with_applicability`) can be used to automatically add the parenthesis if required

Reply:
Thank you, improved!